### ISSUE TITLE
Better Error Handling

### ISSUE THREAD

**Issue by ISNIT0** [ractivejs/ractive#2229]
Not sure if I have missed this somewhere else, but I really would love to see some error handling/events within components.

In our product, if any single component dies (no-matter how nested it may be), the whole page slowly dies as it gets modified.

Even if we could watch for errors in individual components and reload that component, it would be better than some window.onerror = someRefreshFunction;

Any thoughts or things I have missed?

**Comment by heavyk** [comment]
additionally, if a promise gets rejected it will get swallowed. I've found no good way to debug that without finding and replacing the reject code with a debugger statement.

I think whenever I had a component error, it manifested itself as a promise rejection. is that what you're talking abuot?

**Comment by dagnelies** [comment]
What do you mean by component "dying"? If the component produces errors, wouldn't reloading it just produce the same errors again? Then reloading again? Then... ;) ...I'm totally in favor of delivering clear error messages, but extremely suspicious about some automatic reloading.

**Comment by ISNIT0** [comment]
Good point, my issue was fairly vague.

There may be some bug within a component's code that causes something to error.

If that happens, the component itself, and all further components seem to behave oddly, generally different for each individual case, but it seems things like {{#if}} end up not disappearing, some elements get duplicated and so on.

This issue is less about the error reporting, and more allowing for programmatic handling of the errors within Ractive.

Maybe we could get an onerror in a similar way to oninit or oncomplete?

Trying to build a solid example now.

**Comment by fskreuz** [comment]
TLDR: What comprises an error? How do you determine what and where a fatal or non-fatal error?

As far as I remember, Ractive has debug mode on since 0.7. Errors are logged as a warning in the console, caught and *the app continues*. I have observed this in computed properties, which may have funny code or code that depends on something that doesn't exist yet (and returns undefined if it throws up).

Depending on your code, it may or may not cause an invalid state afterwards. Given the computed props example, if a getter throws but it's idempotent (doesn't set state, data etc.) then it's reasonable to have it discontinue and just equate to `undefined`. But if the computed prop happens to set some state and throws up, then that's danger and your app will have invalid state.

So... `onerror` sounds good, but was wondering of when it should be executed and on what :D

In other news, had a related question on error messages with regards to component loaders: https://github.com/ractivejs/ractive/issues/1911 which might be of some help. Loaders need special treatment :trollface:

**Comment by ISNIT0** [comment]
@fskreuz Good point(s), but my definition of an error is a JavaScript error, or anything that turns out as red, triggers window.onerror or is thrown. :)

**Comment by dagnelies** [comment]
Making an example is a wonderful idea ;)

That would clarify the issue and help making ractive more robust ...because its default behavior should indeed be to report the error and continue normally, as good as it can ...and not to break.

**Comment by martypdx** [comment]
Are you looking for same thing as #1919? Components raise `onerror` and `onwarn` events with same info that gets logged to console?

**Comment by martypdx** [comment]
> additionally, if a promise gets rejected it will get swallowed. I've found no good way to debug that without finding and replacing the reject code with a debugger statement.

@heavyk Are you referring to promises returned by the ractive API? Wouldn't you `catch` those yourself?

**Comment by ISNIT0** [comment]
@martypdx Looks about right! Thanks :)

**Comment by Rich-Harris** [comment]
I'll close this in favour of #1919 since they're basically the same
